### PR TITLE
Discover more modules (exes, sub-libraries, test-suites)

### PR DESCRIPTION
**Pull request by ramirez7**
exe `Main` modules are prefixed with the exe name, fixing this old `graphmod` bug: https://github.com/yav/graphmod/issues/28

Fixes https://github.com/dustin/graphex/issues/8

### CODE REVIEW

File: src/Graphex/Cabal.hs [hunk 1]
```diff
@@ -55,14 +56,33 @@ discoverCabalModules cabalFile = do
...
               , path = sourceDirToFilePath srcDir </> Cabal.toFilePath exMod <.> ".hs"
               }
-        ] -- TODO: exes + other-modules
+        , do
+            Executable{..} <- executables
+            srcDir <- hsSourceDirs buildInfo
+            otherMod <- "Main" : buildInfo.otherModules
+            pure Module
+              { name = fromString $
+                if otherMod == "Main"
```
Comment (ramirez7) [inline src/Graphex/Cabal.hs:72]: This is kind of a dumb way to do this, but it works.

File: Makefile [hunk 2]
```diff
@@ -20,3 +20,7 @@ test: graphex.cabal
 fmt:
 	fd .hs --exec stylish-haskell -i
 .PHONY: fmt
+
+graph.json:
+	cabal run -v0 exe:graphex -- cabal > graph.json
```
Comment (ramirez7) [inline Makefile:25]: We need `-v0` due to https://github.com/haskell/cabal/issues/7790

**Comment by ramirez7** [comment]
I want to make discovery command-line configurable + allow the user to optionally also discover external modules as well.

**Comment by ramirez7** [comment]
```
$ graphex cabal --help
Usage: graphex cabal [--discover-exes] [--discover-tests] [--include-external]

  Cabal operations

Available options:
  --discover-exes          Discover exe import dependencies
  --discover-tests         Discover test import dependencies
  --include-external       Include external import dependencies
  -h,--help                Show this help text
```

The only thing left is to add to the existing `graphex cabal` test suite to test the new stuff. I tested with the CLI by hand on `graphex` itself and it seemed correct.

File: src/Graphex/Cabal.hs [hunk 3]
```diff
@@ -54,30 +64,78 @@ sourceDirToFilePath :: FilePath -> FilePath
...
+  , includeExternal :: Bool
+  } deriving stock (Show, Eq)
 
-discoverCabalModuleGraph :: IO ModuleGraph
-discoverCabalModuleGraph = do
+discoverCabalModuleGraph :: CabalDiscoverOpts -> IO ModuleGraph
+discoverCabalModuleGraph opts@CabalDiscoverOpts{..} = do
   fs <- getDirectoryContents "." -- XXX
-  mods <- fmap fold . traverse discoverCabalModules . filter ((".cabal" ==) . takeExtension) $ fs
+  mods <- fmap fold . traverse (discoverCabalModules opts) . filter ((".cabal" ==) . takeExtension) $ fs
```
Comment (dustin) [inline src/Graphex/Cabal.hs:129]: I keep feeling like `fmap fold . traverse` should be a thing.

**Review by dustin — APPROVED** [review]

File: src/Graphex/Cabal.hs (hunk 3 above)
Comment (ramirez7) [reply src/Graphex/Cabal.hs:129]: It's `foldMapA`? But it's one pass:

```haskell
foldMapA
  :: forall t f a b
   . Foldable t
  => Applicative f
  => Monoid b
  => (a -> f b)
  -> t a
  -> f b
foldMapA f = getAp . foldMap (Ap . f)
```

Hm I should probably write a proposal to add it to `base`. I always write it in new projects 😆

**Pull request merged**